pytorch_lightning/loops/fit_loop.py

```diff
@@ -221,27 +228,76 @@ def on_advance_start(self) -> None:  # type: ignore[override]
 
         self.epoch_progress.increment_ready()
 
+        self.trainer.logger_connector.on_epoch_start()
+
+        self.trainer._call_callback_hooks("on_epoch_start")
+        self.trainer._call_lightning_module_hook("on_epoch_start")
+
+        self.trainer._call_callback_hooks("on_train_epoch_start")
+        self.trainer._call_lightning_module_hook("on_train_epoch_start")
+
+        self.epoch_progress.increment_started()
+
     def advance(self) -> None:  # type: ignore[override]
         """Runs one whole epoch."""
         assert self.trainer.train_dataloader is not None
         dataloader = self.trainer.strategy.process_dataloader(self.trainer.train_dataloader)
         data_fetcher = self.trainer._data_connector.get_profiled_dataloader(dataloader)
 
         with self.trainer.profiler.profile("run_training_epoch"):
-            self.epoch_loop.run(data_fetcher)
-
-            # the global step is manually decreased here due to backwards compatibility with existing loggers
-            # as they expect that the same step is used when logging epoch end metrics even when the batch loop has
-            # finished. this means the attribute does not exactly track the number of optimizer steps applied.
-            # TODO(@carmocca): deprecate and rename so users don't get confused
-            self.global_step -= 1
-            # log epoch metrics
-            self.trainer.logger_connector.update_train_epoch_metrics()
-            self.global_step += 1
+            self._outputs = self.epoch_loop.run(data_fetcher)
 
     def on_advance_end(self) -> None:
+        # inform logger the batch loop has finished
+        self.trainer.logger_connector.epoch_end_reached()
+
+        # get the model and call model.training_epoch_end
+        model = self.trainer.lightning_module
+        if is_overridden("training_epoch_end", model) and self._outputs:
```

Comment:
quick ques: what if the user stores the outputs in some state variables within `training_step` and returns only the loss for some reason. Should it avoid calling `training_epoch_end` in such a case too? maybe they could still use `training_epoch_end` hook and clean up their cache manually.

Reply:
There's `on_train_epoch_end` for that. Doing that in `training_epoch_end` would be inefficient because then we save all losses for nothing.

Note also that this condition is copied verbatim from where it was.